internal/commands/build/build.go

```diff
@@ -0,0 +1,233 @@
+	"github.com/moby/buildkit/util/appcontext"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+type buildOptions struct {
+	noCache  bool
+	progress string
+	pull     bool
+	tag      string
+	out      string
+}
+
+func Cmd(dockerCli command.Cli) *cobra.Command {
+	var opts buildOptions
+	cmd := &cobra.Command{
+		Use:     "build [APPLICATION]",
+		Short:   "Build service images for the application",
+		Example: `$ docker app build myapp.dockerapp`,
+		Args:    cli.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			tag, err := runBuild(dockerCli, args[0], opts)
+			if err == nil {
+				fmt.Printf("Successfully build %s\n", tag.String())
+			}
+			return err
+		},
+	}
+
+	flags := cmd.Flags()
+	flags.BoolVar(&opts.noCache, "no-cache", false, "Do not use cache when building the image")
+	flags.StringVar(&opts.progress, "progress", "auto", "Set type of progress output (auto, plain, tty). Use plain to show container output")
+	flags.BoolVar(&opts.pull, "pull", false, "Always attempt to pull a newer version of the image")
+	flags.StringVarP(&opts.out, "output", "o", "", "Dump generated bundle into a file")
+	flags.StringVarP(&opts.tag, "tag", "t", "", "Name and optionally a tag in the 'name:tag' format")
+
+	return cmd
+}
+
+func runBuild(dockerCli command.Cli, application string, opt buildOptions) (reference.Named, error) {
+	ctx := appcontext.Context()
+	info, err := dockerCli.Client().Info(ctx)
+	if err != nil {
+		return nil, err
+	}
+	majorVersion, err := strconv.Atoi(info.ServerVersion[:strings.IndexRune(info.ServerVersion, '.')])
+	if err != nil {
+		return nil, err
+	}
+	if majorVersion < 19 {
+		return nil, errors.New("'build' require docker engine 19.03 or later")
+	}
+
+	app, err := packager.Extract(application)
+	if err != nil {
+		return nil, err
+	}
+	defer app.Cleanup()
+
+	bundle, err := packager.MakeBundleFromApp(dockerCli, app, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	buildopts, err := parseCompose(app, opt)
+	if err != nil {
+		return nil, err
+	}
+
+	buildopts["invocation-image"], err = createInvocationImageBuildOptions(dockerCli, app)
+	if err != nil {
+		return nil, err
+	}
+
+	debugBuildOpts(buildopts)
+
+	d, err := driver.GetDriver(ctx, "buildx_buildkit_default", nil, dockerCli.Client(), nil, "", nil)
+	if err != nil {
+		return nil, err
+	}
+	driverInfo := []build.DriverInfo{
+		{
+			Name:   "default",
+			Driver: d,
+		},
+	}
+
+	ctx2, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	pw := progress.NewPrinter(ctx2, os.Stderr, opt.progress)
+
+	// We rely on buildx "docker" builder integrated in docker engine, so don't nee a DockerAPI here
+	resp, err := build.Build(ctx2, driverInfo, buildopts, nil, dockerCli.ConfigFile(), pw)
+	if err != nil {
+		return nil, err
+	}
+
+	fmt.Println("Successfully built service images")
+	updateBundle(bundle, resp)
+
+	var ref reference.Named
+	ref, err = packager.GetNamedTagged(opt.tag)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Maybe it's something that can be made a the beginning of the function, or at least before the build.
If the tag is not valid, the build will be run and only at the end it will failed. But we can catch it before and warn the user.